python/pyarrow/tests/test_cuda.py

```diff
@@ -313,7 +313,8 @@ def test_CudaBuffer(size):
         assert cbuf[s].to_pybytes() == arr[s].tobytes()
 
     sbuf = cbuf.slice(size//4, size//2)
-    assert sbuf.parent == cbuf
+    cbuf_parent = cuda.CudaBuffer.from_buffer(sbuf.parent)
+    assert cbuf_parent.to_pybytes() == cbuf.to_pybytes()
```

Comment:
What's the reason that this worked before?

Reply:
`sbuf.parent` is pyarrow Buffer with data on GPU and so the `equals` method can not work as we have added a check for non-cpu data.

Why it was working before I am not sure - it segfaulted when testing buffers on GPU for tests in _test_io.py_ 🤷‍♀️

Reply:
Oh, I think I know.

If I compared buffers coming out from the same memory location the `equals` method worked (for example sliced pyarrow Buffer on gpu and sliced CUDA buffer from which the pyarrow Buffer was created, or the example in the cuda test). But if I compared buffers on gpu but on a different memory location, it failed.

Not sure it makes sense - will add a repr (working on it).

Reply:
Looking at `Buffer::Equals`, as you suggested, I understand the behaviour I described in the last comment of this thread. If the addresses of both buffers are equal then `Equals` is true even if they live on GPU. But if the address is different the method uses `memcmp` which will fail for data that is not on CPU.

I added a different check for equals method that checks addresses of buffers that do not live on cpu and raises an error only if they are not the same: https://github.com/apache/arrow/pull/41889/commits/1268c1be2bb863ff300e70e5fbd73047e4e9da5c.

Reply:
> I added a different check for equals method that checks addresses of buffers that do not live on cpu and raises an error only if they are not the same: [1268c1b](https://github.com/apache/arrow/commit/1268c1be2bb863ff300e70e5fbd73047e4e9da5c).

Sounds good. Maybe only add a check that the self and other `.device` is equal as well? (because only at that point comparing the address is fine)